File: ios/MapKmlLayerManager.swift
import GoogleMaps
import GoogleMapsUtils

private final class KmlLayerState {
  let current: RNKMLayer
  var renderer: GMUGeometryRenderer?

  init(current: RNKMLayer) {
    self.current = current
  }
}

final class MapKmlLayerManager {
  private let mapErrorHandler: MapErrorHandler
  private weak var mapView: GMSMapView?
  private var states: [String: KmlLayerState] = [:]
  private var destroyed = false

  init(mapErrorHandler: MapErrorHandler) {
    self.mapErrorHandler = mapErrorHandler
  }

  func attachMap(_ mapView: GMSMapView) {
    onMain {
      guard !self.destroyed else { return }
      self.mapView = mapView
      self.states.values.filter { $0.renderer == nil }.forEach { self.addToMap($0) }
    }
  }

  func add(_ kmlLayer: RNKMLayer) {
    onMain {
      guard !self.destroyed else { return }
      self.remove(id: kmlLayer.id)
      let state = KmlLayerState(current: kmlLayer)
      self.states[kmlLayer.id] = state
      if self.mapView != nil {
        self.addToMap(state)
      }
    }
  }

  func update(_ next: RNKMLayer) {
    onMain {
      guard !self.destroyed else { return }
      guard let state = self.states[next.id] else { return }
      if state.current.kmlLayerEquals(next) { return }
      self.add(next)
    }
  }

  func remove(id: String) {
    onMain {
      self.states.removeValue(forKey: id).map { self.removeFromMap($0) }
    }
  }

  func destroy() {
    onMain {
      self.destroyed = true
      self.states.values.forEach { self.removeFromMap($0) }
      self.states.removeAll()
      self.mapView = nil
    }
  }

  private func addToMap(_ state: KmlLayerState) {
    guard let mapView = self.mapView else { return }
    guard let data = state.current.kmlString.data(using: .utf8) else {
      self.mapErrorHandler.report(RNMapErrorCode.kmlLayerFailed, "kmlLayerId=\(state.current.id) parse failed")
      return
    }
    let parser = GMUKMLParser(data: data)
    parser.parse()
    let renderer = GMUGeometryRenderer(
      map: mapView,
      geometries: parser.placemarks,
      styles: parser.styles,
      styleMaps: parser.styleMaps
    )
    renderer.render()
    state.renderer = renderer
  }

  private func removeFromMap(_ state: KmlLayerState) {
    state.renderer?.clear()
  }
}

File: android/src/main/java/com/rngooglemapsplus/MapKmlLayerManager.kt
package com.rngooglemapsplus

import android.content.Context
import com.google.android.gms.maps.GoogleMap
import com.google.maps.android.data.kml.KmlLayer
import com.rngooglemapsplus.extensions.kmlLayerEquals
import java.io.ByteArrayInputStream
import java.nio.charset.StandardCharsets

private class KmlLayerState(
  val current: RNKMLayer,
) {
  var layer: KmlLayer? = null
}

class MapKmlLayerManager(
  private val context: Context,
  private val mapErrorHandler: MapErrorHandler,
) {
  private var map: GoogleMap? = null
  private val states = mutableMapOf<String, KmlLayerState>()
  private var destroyed = false

  fun attachMap(map: GoogleMap) =
    onUi {
      if (destroyed) return@onUi
      this.map = map
      states.values.filter { it.layer == null }.forEach { addToMap(it) }
    }

  fun add(kmlLayer: RNKMLayer) =
    onUi {
      if (destroyed) return@onUi
      remove(kmlLayer.id)
      val state = KmlLayerState(kmlLayer)
      states[kmlLayer.id] = state
      if (map != null) addToMap(state)
    }

  fun update(next: RNKMLayer) =
    onUi {
      if (destroyed) return@onUi
      val state = states[next.id] ?: return@onUi
      if (state.current.kmlLayerEquals(next)) return@onUi
      add(next)
    }

  fun remove(id: String) =
    onUi {
      states.remove(id)?.let { removeFromMap(it) }
    }

  fun destroy() =
    onUi {
      destroyed = true
      states.values.forEach { removeFromMap(it) }
      states.clear()
      map = null
    }

  private fun addToMap(state: KmlLayerState) {
    val map = map ?: return
    try {
      val inputStream = ByteArrayInputStream(state.current.kmlString.toByteArray(StandardCharsets.UTF_8))
      val layer = KmlLayer(map, inputStream, context)
      state.layer = layer
      layer.addLayerToMap()
    } catch (e: Exception) {
      mapErrorHandler.report(RNMapErrorCode.KML_LAYER_FAILED, "kmlLayerId=${state.current.id} parse failed", e)
    }
  }

  private fun removeFromMap(state: KmlLayerState) {
    state.layer?.removeLayerFromMap()
  }
}
